### DISCUSSION TITLE
Connecting Isolated Pores and Labeling after trimming

### DISCUSSION THREAD

**Discussion by Fioti38** [PMEAL/OpenPNM#2395] [Q&A] [answered]
Hello , 1st thank you for this amazing tool you're providing.

I'm looking for a way to connect pores artificially to simulate the flow that is moving in undetectable pores .

 For this, I found this function in your documentation "connect_pores(network, pores1, pores2, labels=[], add_conns=True)",
but I have no idea how to define pores1 and pores2, because I don't find how to render pores positions (x,y,z) to get pore indexes to be connected. 
Is it possible to apply a connect_pores on all pores network with a defined throat geometry(e.g small diameter)?

I'm also having a problem to update labels after trimming, labels (defined when extracting the network from tiff image)such as xmin xmax zmin zmax ymin ymax aren't updated and I can't apply my boundary conditions to simulate the flow. (e.g: these labels are empty after trimming)

I'm working on a XCT tiff image , pores network extraction is done by porespy snow2 algorithm.
Thank you in advance

**Comment by Fioti38** [comment, marked as answer]
Any hints on how to do this please ?

**Reply by jgostick** [reply]
oops, sorry this slipped through the cracks.  

So, if I understand correctly, you would like to connect pores that are not direct neighbors, but are close-ish to each other.  There is a method called ``find_nearby_pores`` which you can use to generate candidate pores, such that pore 1 should be connected to 5, 33, etc.  Once you have this you can use the ``extend`` method to add new throats between these pores.  The extend method would work by passing a list of new connections to make, like [[1, 5], [1, 33], [1, etc]].  You'll have to find which pores you want to connect yourself if the "nearby" approach does not work.  You'll also have to compute the hydraulic and diffusive conductance of these pores yourself, which will require add a new geometry and physics object to the project.[^1]

[^1]: In version 3, which we are working on very hard these days, the concept of geometry and physics will be gone, so if you find yourself feeling very frustrated by this aspect, then you can at least know that we're trying to fix it.

**Reply by Fioti38** [reply]
Thank you for your answer, so if I would like to make a connections list as you said [pore 1, pore 5] etc ... I need the index of the pores I want to connect right? is there any way I could render a list of all pores with their X,Y,Z positions so that I can recognize the pores indexes I want to connect ?

**Reply by jgostick** [reply]
You can view the coordinates of each pore with ``net['pore.coords']``.  You can do thresholds on this like ``hits = pn['pore.coords'][:, 0] < val`` to get a boolean mask of all pores with an x-coords less than val.  You can then use ``np.where(hits)`` to get the index of the ``True`` values. Note that ``np.where`` returns a tuple, so you'll need to index into it to get a list of pore indices (i.e. ``np.where(hits)[0]``.  How that helps.

**Comment by Fioti38** [comment]
Thank you so much for your answer, this really helped I also found out easy to connect clustered pores with stitch pores function after using find_nerby_pores with a defined radius, but when I do a single phase flow simulation on this new network, im getting this error that I couldn't understand 
  _File "...site-packages\openpnm\models\physics\hydraulic_conductance.py", line 55, in generic_hydraulic
    gt = F[f"{size_factors}.throat"][throats] / mut

ValueError: operands could not be broadcast together with shapes (982,) (2323,)_ 

Merci

**Reply by jgostick** [reply]
Did you do your stitching after creating the 'other' objects, like geometry and physics?  That would explain the mis-matched shapes.  If you uou do it before creating any other objects, then all subsequent objects will be the correct size.  The extend method is supposed do this for you, but it can be a bit fragile and subject to "edge cases".

**Reply by Fioti38** [reply]
Thank you for you quick answer , actually this is my script in this case : 
![image](https://user-images.githubusercontent.com/104492138/168038051-a128e598-f540-4eee-8ea0-488a314b6389.png)

I'm still getting this error ValueError: operands could not be broadcast together with shapes.

Another question regarding when I import a network from porespy, is there a way to save this snow.network to a file that I can directly load, to avoid losing time in snow extraction everytime I run my code 

Thank you so much for the value you are bringing to the research community.

**Reply by jgostick** [reply]
from the snippet above, it looks like you must have a ``geo1`` since you are defining a ``geo2``.    Are you able to do the connect_pores step before creating any Geometries?  Can you try it just as a test?

**Reply by jgostick** [reply]
Regarding your second question, yes, you can use the Python ``pickle`` module:

``` python
import pickle as pk
with open('output.dat', 'wb') as f:
    pk.dump(snow, f)

```
which you can then open with:

``` pyhon
snow = pk.load(open('output.dat', 'rb'))
```

**Reply by Fioti38** [reply]
> from the snippet above, it looks like you must have a `geo1` since you are defining a `geo2`. Are you able to do the connect_pores step before creating any Geometries? Can you try it just as a test?
Thank you so much for your reeactivity and your answer,
Yes , I have a geo1 defined before from an imported network, so  Im defining a geo2 for my added throats 

Yes I'm able to do connect_pores and generate additional connections (In my geo from extraction I have 14764 throats, after connecting pores I have 2027402 throats ), so do I need to define diameter and length for these new throats ? because when checking these throats properties the table is empty. and the error I'm getting looks like it has some relation with hydraulic conductivity : 
The error i'm getting : 
  File "...\lib\site-packages\openpnm\models\physics\hydraulic_conductance.py", line 55, in generic_hydraulic
    gt = F[f"{size_factors}.throat"][throats] / mut
ValueError: operands could not be broadcast together with shapes (9014,) (2027402,)

**Reply by jgostick** [reply]
Yes, you must add properties for these new connections.  Can possibly post your full script, at least up until the trouble begins, then I'll see if I can make it work?

**Reply by Fioti38** [reply]
So there is a way to add a general dimeter to all throats and let it compute the hydraulic conductivity ?
Ok thank you so much , this is my full script :

``` python
import numpy as np
import openpnm as op

import tifffile
import pickle as pk

im = tifffile.imread("pores-pnm.tif") 
im = ~np.array(im, dtype=bool)[:, :, :]
np.set_printoptions(precision=4)
np.random.seed(10)

resolution = 104e-6

    
snow = pk.load(open('output.dat', 'rb'))
   
pn, geo = op.io.PoreSpy.import_data(snow.network)

h = pn.check_network_health()

Ps2=pn.find_nearby_pores(pores=h['trim_pores'], r=0.5)

Ps1=h['trim_pores']
op.topotools.connect_pores(pn, Ps1, Ps2[0], labels='new', add_conns=True)
geo2 = op.geometry.GenericGeometry(network=pn,throats=pn.throats('new'))


h = pn.check_network_health()
op.topotools.trim(network=pn, pores=h['trim_pores'])

print(op.topotools.is_fully_connected(pn))

#assign physics
water = op.phases.Water(network=pn)

phys = op.physics.Basic(network=pn, phase=water, geometry=geo)
phy2 = op.physics.Basic(network=pn, phase=water, geometry=geo2)

Pin=2000000
Pout=1000000

perm = op.algorithms.StokesFlow(network=pn, phase=water)
perm.set_value_BC(pores=pn['pore.xmax'], values=Pout)
perm.set_value_BC(pores=pn['pore.xmin'], values=Pin)
perm.run()

water.update(perm.results())



Q = perm.rate(pores=pn.pores('xmin'), mode='group')[0]
print(Q)

A = (im.shape[1] * im.shape[2]) * resolution**2
L = im.shape[0] * resolution
mu = water['pore.viscosity'].max()
print(mu)
delta_P = Pin - Pout

K = Q * L * mu / (A * delta_P)

print(f'The value of K is: {K} m2')
```

**Reply by Fioti38** [reply]
Hello , 
Were able to take a look at this piece of code? I tried to add manually the throat hydraulic conductance for the new added throats 
phy2['throat.hydraulic_conductance'] = 1.0*resolution**2
but it didn't make any difference
Thank you again so much

**Reply by jgostick** [reply]
sorry, I'm swamped at the moment.

**Reply by Fioti38** [reply]
Of course I understand, in the meanwhile I think that to run this code you need the ouput.dat file which has the network data, here is a link where you can download it : https://enacshare.epfl.ch/f5A2bPj7k9VuLHGCvE3NQtSanZhxygf

Thank you so much in advance

**Reply by jgostick** [reply]
what happens if you change this line:

```
geo2 = op.geometry.GenericGeometry(network=pn, throats=pn.throats('new'))
```

to this:

```
geo2 = op.geometry.SpheresAndCylinders(network=pn, throats=pn.throats('new'))
```

The point is that you have geometrical information for all the pores and throats from the extraction stored on ``geo``, but no geometrical properties for the new throats that you add.  The ``StickAndBall`` class has pore-scale models attached that will compute geometrical values for the throats which it owns.  It will also calculate throat length, volume, etc.  This *should* then make the rest of the simulation run smoothly since all the pores and throats will have a full set of geometrical values.

**Reply by Fioti38** [reply]
Thank you so much for your answer,
 I've changed the geometry model as you suggested and this leads to get geometrical values, but this didn't solve this problem which I think is linked to the difference between shapes of the phase _(2027402,)_ and number of throats from geo1 _(9014,)_, so in _hydraulic_conductance.py_ when computing `gt = F[f"{size_factors}.throat"][throats] / mut `, this leads to an error . So, for example if there is a way to either merge two geometries in one? or when defining physics with something like this : `phys = op.physics.Basic(network=pn, phase=water,geometry=geo+geo2)` ?

Thank you

**Reply by jgostick** [reply]
Can you do ``print(obj)`` for all the objects and post?

**Reply by Fioti38** [reply]
Here is a print for pn, 2 geos and phase
```
openpnm.network.GenericNetwork : net_01
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Properties                                    Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.coords                                   14764 / 14764
2     pore.region_label                             14764 / 14764
3     throat.conns                                   9014 / 9014 
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      14764     
2     pore.boundary                                 6         
3     pore.geo_01                                   14764     
4     pore.xmax                                     1         
5     pore.xmin                                     1         
6     pore.ymax                                     1         
7     pore.ymin                                     1         
8     pore.zmax                                     1         
9     pore.zmin                                     1         
10    throat.all                                    9014      
11    throat.geo_01                                 9014      
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
openpnm.geometry.Imported : geo_01
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Properties                                    Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.diameter                                 14764 / 14764
2     pore.equivalent_diameter                      14764 / 14764
3     pore.extended_diameter                        14764 / 14764
4     pore.geometric_centroid                       14764 / 14764
5     pore.global_peak                              14764 / 14764
6     pore.inscribed_diameter                       14764 / 14764
7     pore.local_peak                               14764 / 14764
8     pore.phase                                    14764 / 14764
9     pore.region_volume                            14764 / 14764
10    pore.surface_area                             14764 / 14764
11    pore.volume                                   14764 / 14764
12    throat.cross_sectional_area                    9014 / 9014 
13    throat.diameter                                9014 / 9014 
14    throat.diffusive_size_factors.pore1            9014 / 9014 
15    throat.diffusive_size_factors.pore2            9014 / 9014 
16    throat.diffusive_size_factors.th...            9014 / 9014 
17    throat.direct_length                           9014 / 9014 
18    throat.equivalent_diameter                     9014 / 9014 
19    throat.global_peak                             9014 / 9014 
20    throat.hydraulic_size_factors.pore1            9014 / 9014 
21    throat.hydraulic_size_factors.pore2            9014 / 9014 
22    throat.hydraulic_size_factors.th...            9014 / 9014 
23    throat.inscribed_diameter                      9014 / 9014 
24    throat.length                                  9014 / 9014 
25    throat.perimeter                               9014 / 9014 
26    throat.phases                                  9014 / 9014 
27    throat.total_length                            9014 / 9014 
28    throat.volume                                  9014 / 9014 
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      14764     
2     throat.all                                    9014      
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
――――
openpnm.geometry.SpheresAndCylinders : geo_02
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Properties                                    Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.diameter                                     0 / 0    
2     pore.max_size                                     0 / 0    
3     pore.seed                                         0 / 0    
4     pore.volume                                       0 / 0    
5     throat.cross_sectional_area                   2018388 / 2018388
6     throat.diameter                               2018388 / 2018388
7     throat.diffusive_size_factors.pore1           2018388 / 2018388
8     throat.diffusive_size_factors.pore2           2018388 / 2018388
9     throat.diffusive_size_factors.th...           2018388 / 2018388
10    throat.hydraulic_size_factors.pore1           2018388 / 2018388
11    throat.hydraulic_size_factors.pore2           2018388 / 2018388
12    throat.hydraulic_size_factors.th...           2018388 / 2018388
13    throat.length                                 2018388 / 2018388
14    throat.max_size                               2018388 / 2018388
15    throat.volume                                 2018388 / 2018388
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      0         
2     throat.all                                    2018388   
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
openpnm.phases.Water : phase_01
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Properties                                    Valid Values
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.contact_angle                            14764 / 14764
2     pore.critical_pressure                        14764 / 14764
3     pore.critical_temperature                     14764 / 14764
4     pore.critical_volume                          14764 / 14764
5     pore.density                                  14764 / 14764
6     pore.diffusivity                              14764 / 14764
7     pore.electrical_conductivity                  14764 / 14764
8     pore.molar_density                            14764 / 14764
9     pore.molecular_weight                         14764 / 14764
10    pore.pressure                                 14764 / 14764
11    pore.surface_tension                          14764 / 14764
12    pore.temperature                              14764 / 14764
13    pore.thermal_conductivity                     14764 / 14764
14    pore.vapor_pressure                           14764 / 14764
15    pore.viscosity                                14764 / 14764
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
#     Labels                                        Assigned Locations
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
1     pore.all                                      14764     
2     throat.all                                    2027402   
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```
